drivers/gps/nrf9160_gps/nrf9160_gps.c

```diff
@@ -13,87 +13,140 @@
+ * event item because an NMEA string would get overwritten by the next NMEA string before the
+ * consumer thread is able to read it. This is not a problem with PVT and A-GPS request data, but
+ * all data is handled in the same way for simplicity.
+ */
+struct event_item {
+	uint8_t id;
+	void    *data;
 };
 
-static int stop_gps(const struct device *dev, bool is_timeout);
+K_MSGQ_DEFINE(event_msgq, sizeof(struct event_item), 10, 4);
 
 static uint64_t start_timestamp;
 static uint64_t fix_timestamp;
 
-static nrf_gnss_agps_data_type_t type_lookup_gps2socket[] = {
-	[GPS_AGPS_UTC_PARAMETERS]	= NRF_GNSS_AGPS_UTC_PARAMETERS,
-	[GPS_AGPS_EPHEMERIDES]		= NRF_GNSS_AGPS_EPHEMERIDES,
-	[GPS_AGPS_ALMANAC]		= NRF_GNSS_AGPS_ALMANAC,
+static uint16_t type_lookup_gps2gnss_api[] = {
+	[GPS_AGPS_UTC_PARAMETERS]	= NRF_MODEM_GNSS_AGPS_UTC_PARAMETERS,
+	[GPS_AGPS_EPHEMERIDES]		= NRF_MODEM_GNSS_AGPS_EPHEMERIDES,
+	[GPS_AGPS_ALMANAC]		= NRF_MODEM_GNSS_AGPS_ALMANAC,
 	[GPS_AGPS_KLOBUCHAR_CORRECTION]
-		= NRF_GNSS_AGPS_KLOBUCHAR_IONOSPHERIC_CORRECTION,
+		= NRF_MODEM_GNSS_AGPS_KLOBUCHAR_IONOSPHERIC_CORRECTION,
 	[GPS_AGPS_NEQUICK_CORRECTION]
-		= NRF_GNSS_AGPS_NEQUICK_IONOSPHERIC_CORRECTION,
+		= NRF_MODEM_GNSS_AGPS_NEQUICK_IONOSPHERIC_CORRECTION,
 	[GPS_AGPS_GPS_SYSTEM_CLOCK_AND_TOWS]
-		= NRF_GNSS_AGPS_GPS_SYSTEM_CLOCK_AND_TOWS,
-	[GPS_AGPS_LOCATION]		= NRF_GNSS_AGPS_LOCATION,
-	[GPS_AGPS_INTEGRITY]		= NRF_GNSS_AGPS_INTEGRITY,
+		= NRF_MODEM_GNSS_AGPS_GPS_SYSTEM_CLOCK_AND_TOWS,
+	[GPS_AGPS_LOCATION]		= NRF_MODEM_GNSS_AGPS_LOCATION,
+	[GPS_AGPS_INTEGRITY]		= NRF_MODEM_GNSS_AGPS_INTEGRITY,
 };
 
-static void copy_pvt(struct gps_pvt *dest, nrf_gnss_pvt_data_frame_t *src)
+static int get_event_data(void **dest, uint8_t type, size_t len)
+{
+	void *data;
+
+	data = k_malloc(len);
+	if (data == NULL) {
+		return -1;
+	}
+
+	if (nrf_modem_gnss_read(data, len, type) != 0) {
+		k_free(data);
+		return -1;
+	}
+
+	*dest = data;
+
+	return 0;
+}
+
+/* This handler is called in ISR, so need to avoid heavy processing. */
+static void gnss_event_handler(int event_id)
+{
+	int err = 0;
+	struct event_item event;
+
+	event.id = event_id;
+	event.data = NULL;
+
+	switch (event_id) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		err = get_event_data(&event.data,
+				     NRF_MODEM_GNSS_DATA_PVT,
+				     sizeof(struct nrf_modem_gnss_pvt_data_frame));
+		break;
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		err = get_event_data(&event.data,
+				     NRF_MODEM_GNSS_DATA_NMEA,
+				     sizeof(struct nrf_modem_gnss_nmea_data_frame));
+		break;
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+		err = get_event_data(&event.data,
+				     NRF_MODEM_GNSS_DATA_AGPS_REQ,
+				     sizeof(struct nrf_modem_gnss_agps_data_frame));
+		break;
+	default:
+		/* All other events are just ignored. */
+		return;
+	}
+
+	if (err) {
+		/* Failed to get event data. */
+		return;
+	}
+
+	err = k_msgq_put(&event_msgq, &event, K_NO_WAIT);
+	if (err) {
+		/* Failed to put event into queue. */
```

Comment:
If this API fails it might be due to the message queue being filled which would indicate that the processing thread might be blocked somehow. The application should probably be notified with an error and/or the message queue should be purged.

Reply:
Are you thinking about a warning/error trace or something else? I think a trace would be nice, but gnss_event_handler() is called from an ISR, so I'm not sure we want to use Logger? If deferred logging is not used, it might cause problems. Would printk() be better?

Reply:
I just wonder if we can afford not processing events received via the GNSS API when they indirectly translate to some state changing event in the application. (GPS_EVT_START, FIX, TIMEOUT). If we can get into an undefined state in the application and that we should consider a filled message queue in this situation a fatal error really. `k_msgq_put` should never really fail and if it does we should detect it as early as possible. Maybe a run-time assert is the way to go? Or callback to the application (GPS_EVT_ERROR with an error cause). So that we atleast trasfer some awareness of the problem that occurred.

Reply:
I get your point. An assert sounds like a good idea to me, that makes the issue visible immediately. If the message queue becomes full, the application might for example miss a fix, so the code should be changed so that it can process all events fast enough.

Reply:
As discussed offline, I'll add both an assert and an error event in case asserts are disabled.

Reply:
Assert caused an assert loop, probably because the handler is being run in ISR. Because of this I only added sending of GPS_EVT_ERROR event. I also changed the implementation in asset_tracker_v2 so that GPS_EVT_ERROR is handled as a fatal error.